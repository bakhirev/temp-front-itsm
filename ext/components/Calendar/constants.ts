export type LocaleType = 'ru' | 'enUS' | 'de';

export type Corners = {
  [key in 'top-left' | 'bottom-left' | 'top-right' | 'bottom-right']?: boolean;
};

export type CalendarSize = 'big' | 'small';

export const SIZES = {
  big: 320,
  small: 248,
};

export const NAVIGATION_PANEL_SIZES = {
  big: 64,
  small: 52,
};

export const DAY_SIZES = {
  big: 40,
  small: 32,
};

export const DAY_PADDINGS = {
  big: 12,
  small: 8,
};

export const NAVIGATION_YEAR_SIZES = {
  big: 8,
  small: 6,
};

export const YEAR_SIZES = {
  big: 68,
  small: 56,
};

export const YEAR_PADDINGS = {
  big: 26,
  small: 20,
};

export const DAY_VIEW_PADDINGS = {
  big: '28px 20px 16px',
  small: '20px 12px 12px',
};

export const YEAR_VIEW_PADDINGS = {
  big: '28px 24px 16px',
  small: '20px 12px 12px',
};

export const HEADER_PADDINGS = {
  big: '26px 0px 28px 32px',
  small: '20px 0px 20px 20px',
};

export const PANEL_DAY_VIEW_PADDINGS = {
  big: '0 4px 28px 10px',
  small: '0 4px 20px 8px',
};

export const PANEL_YEAR_VIEW_PADDINGS = {
  big: '0 0 28px 6px',
  small: '0 4px 20px 8px',
};

export const DAY_NAMES_BOTTOM_MARGINS = {
  big: 12,
  small: 8,
};

export const DAY_NUMBERS = [0, 1, 2, 3, 4, 5, 6];

export const MENU_DAY = 'Выбор дня';
export const MENU_YEAR = 'Выбор года';

export const MENU_NEXT_YEAR = 'Вперед';
export const MENU_PREVIOUS_YEAR = 'Назад';

export const MENU_NEXT_MONTH = 'Следующий месяц';
export const MENU_PREVIOUS_MONTH = 'Предыдущий месяц';

export const DEFAULT_SIZE = 'big';
export const DEFAULT_YEAR_COUNT = 16;
export const DEFAULT_LOCALE_NAME = 'ru';
