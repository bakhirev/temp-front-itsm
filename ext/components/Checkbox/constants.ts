export type CheckboxSize = 'big' | 'small';

export const CHECKBOX_MIN_HEIGHT_SMALL = '20px';
export const CHECKBOX_MIN_HEIGHT_BIG = '24px';
export const MARK_BEFORE_OFFSET = '-9px';
export const MARK_HEIGHT_SMALL = '16px';
export const MARK_HEIGHT_BIG = '20px';
export const MARK_WIDTH_SMALL = '16px';
export const MARK_WIDTH_BIG = '20px';
export const CHECKBOX_MARGIN = '2px';
export const CHECKBOX_MARGIN_WITH_LABEL = '10px';
export const INDETERMINATE_SVG_WIDTH_SMALL = '8px';
export const INDETERMINATE_SVG_WIDTH_BIG = '10px';
export const BASIC_SVG_WIDTH_SMALL = '10px';
export const BASIC_SVG_WIDTH_BIG = '12px';
