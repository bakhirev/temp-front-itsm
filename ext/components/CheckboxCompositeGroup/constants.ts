export const ITEM_PADDING_LEFT_SMALL = '28px';
export const ITEM_PADDING_LEFT_BIG = '32px';
export const ITEM_MARGIN_BOTTOM_SMALL = '12px';
export const ITEM_MARGIN_BOTTOM_BIG = '16px';
export const ITEM_BEFORE_WIDTH_SMALL = '18px';
export const ITEM_BEFORE_WIDTH_BIG = '20px';
export const ITEM_BEFORE_LEFT_SMALL = '10px';
export const ITEM_BEFORE_LEFT_BIG = '12px';
export const COMPOSITEGROUP_PADDING_TOP_SMALL = '12px';
export const COMPOSITEGROUP_PADDING_TOP_BIG = '16px';
export const COMPOSITEGROUP_BEFORE_HEIGHT_SMALL = '12px';
export const COMPOSITEGROUP_BEFORE_HEIGHT_BIG = '14px';
export const COMPOSITEGROUP_BEFORE_LEFT_SMALL = '10px';
export const COMPOSITEGROUP_BEFORE_LEFT_BIG = '12px';
export const COMPOSITEGROUP_BEFORE_TOP = '2px';
