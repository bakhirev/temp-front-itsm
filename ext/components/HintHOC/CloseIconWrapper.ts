import styled, { css } from 'styled-components';

import { DEFAULT_THEME, markerStyle } from '../common';

const MARKER_OFFSET = 6;
const RIGHT = 16;
const TOP = 12;

export const CloseIconWrapper = styled.div<{ light?: boolean }>`
  ${({ light, theme }) => css`
    align-items: center;
    display: flex;
    position: absolute;
    right: ${RIGHT}px;
    top: ${TOP}px;
    ${markerStyle(theme, MARKER_OFFSET, !light)};

    &:hover {
      cursor: pointer;
    }

    & svg {
      fill: ${light ? theme.color.neutral[50] : theme.color.neutral.white};
    }
  `};
`;

CloseIconWrapper.defaultProps = {
  theme: DEFAULT_THEME,
};
