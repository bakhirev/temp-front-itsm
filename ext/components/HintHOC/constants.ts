export type Align = 'left' | 'center' | 'right';
export type Direction = 'bottom' | 'top';
export type Size = 'big' | 'medium' | 'small';

export const SIZES: { [key in Size]: number } = {
  big: 488,
  medium: 384,
  small: 280,
};

export const POINTER_HEIGHT = 8;
export const POINTER_MARGIN = 18;
export const POINTER_WIDTH = 16;

export const SAFE_SPACE = 16;
