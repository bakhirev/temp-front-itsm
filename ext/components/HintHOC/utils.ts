import { css } from 'styled-components';

import { POINTER_HEIGHT, POINTER_MARGIN, POINTER_WIDTH, SAFE_SPACE } from './constants';
import type { Align, Direction } from './constants';

const HINT_MARGIN = 4;

interface ISize {
  width: number;
  height: number;
}

interface IRect extends ISize {
  left: number;
  right: number;
  top: number;
  bottom: number;
}

export const getPointerDirectionStyle = (direction: Direction, height: number) => {
  switch (direction) {
    case 'top':
      return css`
        bottom: -${height}px;
        transform: rotate(180deg);
      `;
    case 'bottom':
    default:
      return css`
        top: -${height}px;
      `;
  }
};

export const getPointerAlignStyle = (align: Align, width: number) => {
  switch (align) {
    case 'left':
      return css`
        right: ${POINTER_MARGIN}px;
      `;
    case 'center':
      return css`
        left: calc(50% - (${width}px / 2));
      `;
    case 'right':
    default:
      return css`
        left: ${POINTER_MARGIN}px;
      `;
  }
};

export const findDirection = (componentRect: IRect, targetRect: IRect) => {
  const viewPortHeight = window.innerHeight;
  if (viewPortHeight < targetRect.bottom + HINT_MARGIN + componentRect.height + SAFE_SPACE) {
    return 'top';
  }
  return 'bottom';
};

export const findAlign = (componentRect: IRect, targetRect: IRect): Align => {
  const viewPortWidth = window.innerWidth;
  const componentHorizontalAvailableSpace =
    componentRect.width - (targetRect.width + POINTER_WIDTH) / 2 - POINTER_MARGIN + SAFE_SPACE;
  if (targetRect.left >= componentHorizontalAvailableSpace) {
    return 'left';
  }
  if (viewPortWidth - targetRect.left >= componentHorizontalAvailableSpace) {
    return 'right';
  }
  return 'center';
};

export const calculatePosition = (
  direction: Direction,
  align: Align,
  componentSize: ISize,
  targetSize: ISize,
  offset: ISize
) => {
  let posX: number, posY: number;

  switch (direction) {
    case 'top':
      posY = -(componentSize.height + POINTER_HEIGHT + HINT_MARGIN);
      break;
    case 'bottom':
    default:
      posY = targetSize.height + POINTER_HEIGHT + HINT_MARGIN;
      break;
  }

  switch (align) {
    case 'left':
      posX = (targetSize.width + POINTER_WIDTH) / 2 + POINTER_MARGIN - componentSize.width;
      break;
    case 'center':
      posX = (targetSize.width - componentSize.width) / 2;
      break;
    case 'right':
    default:
      posX = (targetSize.width - POINTER_WIDTH) / 2 - POINTER_MARGIN;
      break;
  }
  return { posX: posX + offset.width, posY: posY + offset.height };
};
