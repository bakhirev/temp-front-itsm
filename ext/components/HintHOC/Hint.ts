import styled, { css } from 'styled-components';

import { DEFAULT_THEME, Z_INDEXES } from '../common';

import { POINTER_HEIGHT, POINTER_WIDTH } from './constants';
import { getPointerAlignStyle, getPointerDirectionStyle } from './utils';
import type { Align, Direction } from './constants';

const PADDING = '12px 52px 12px 16px';

interface IHint {
  light?: boolean;
  offset?: number;
  pointerAlign: Align;
  pointerDirection: Direction;
  position: { posX: number; posY: number };
  visible?: boolean;
  width: number;
}

export const Hint = styled.div<IHint>`
  ${({
    light,
    offset,
    pointerAlign,
    pointerDirection,
    position: { posX, posY },
    theme,
    visible,
    width,
  }) => css`
    align-items: center;
    background: ${light ? theme.color.neutral.white : theme.color.neutral[80]};
    border-radius: ${theme.borderRadius};
    box-shadow: ${light ? theme.boxShadow[12] : 'none'};
    box-sizing: border-box;
    color: ${light ? theme.color.neutral[90] : theme.color.neutral.white};
    display: flex;
    left: ${posX}px;
    padding: ${PADDING};
    position: fixed;
    top: ${offset ? posY + offset : posY}px;
    visibility: ${visible ? 'visible' : 'hidden'};
    width: ${width}px;
    z-index: ${Z_INDEXES.HINT};

    &::after {
      border-color: transparent transparent
        ${light ? theme.color.neutral.white : theme.color.neutral[80]} transparent;
      border-style: solid;
      border-width: 0 ${POINTER_HEIGHT}px ${POINTER_HEIGHT}px ${POINTER_HEIGHT}px;
      content: '';
      height: 0;
      position: absolute;
      width: 0;
      ${getPointerAlignStyle(pointerAlign, POINTER_WIDTH)};
      ${getPointerDirectionStyle(pointerDirection, POINTER_HEIGHT)};
    }
  `};
`;

Hint.defaultProps = {
  theme: DEFAULT_THEME,
};
