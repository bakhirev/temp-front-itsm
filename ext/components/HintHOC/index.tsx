import React, { createRef, PureComponent } from 'react';
import ReactDOM from 'react-dom';
import type { ComponentType, RefObject } from 'react';

import { Body2Long } from '../Typography';
import { containNode, getContainingBlockOffset, getScrollableParents } from '../common/utils';
import { ReactComponent as CloseOutline } from '../Icons/service/CloseOutline.svg';

import { CloseIconWrapper } from './CloseIconWrapper';
import { Hint } from './Hint';
import { Wrapper } from './Wrapper';
import { calculatePosition, findAlign, findDirection } from './utils';
import { SIZES } from './constants';
import type { Align, Direction, Size } from './constants';

export interface IHintHOCProps {
  /** Выравнивание подсказки */
  align?: Align;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** Контейнер, в котором происходит размещение.
   * В отличие от TooltipHOC, данный компонент по дизайну всегда имеет target, относительно
   * которого будет происходить позиционирование - это оборачиваемый компонент (иконка и т.п.) */
  container?: Element | null;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Направление подсказки */
  direction?: Direction;
  /** Светлый вариант подсказки */
  light?: boolean;
  /** Текст подсказки */
  message: string;
  /** Отступ подсказки */
  offset?: number;
  /** Размер подсказки */
  size?: Size;
  /** Показать/скрыть подсказку */
  visible?: boolean;
  /** Обработчик, срабатывающий при попытке пользователя закрыть подсказку */
  onRequestHide?: () => void;
}

interface IHintHOCState {
  pointerAlign: Align;
  pointerDirection: Direction;
  posX: number;
  posY: number;
}

const DEFAULT_POINTER_ALIGN: Align = 'left';
const DEFAULT_POINTER_DIRECTION: Direction = 'bottom';

export const createHintHOC = <P extends Object>(WrappedComponent: ComponentType<P>) =>
  class HintHOC extends PureComponent<P & IHintHOCProps, IHintHOCState> {
    hintRef: RefObject<HTMLDivElement>;
    wrapperRef: RefObject<HTMLDivElement>;

    scrollableParents?: Array<Element> = undefined;

    constructor(props: P & IHintHOCProps) {
      super(props);
      this.hintRef = createRef();
      this.wrapperRef = createRef();

      this.state = {
        pointerAlign: DEFAULT_POINTER_ALIGN,
        pointerDirection: DEFAULT_POINTER_DIRECTION,
        posX: 0,
        posY: 0,
      };
    }

    componentDidMount() {
      document.addEventListener('click', this.handleDocumentClick.bind(this));
      window.addEventListener('scroll', this.onHide);
      window.addEventListener('resize', this.onHide);
      this.changePosition();
    }

    componentDidUpdate(prevProps: IHintHOCProps, _: IHintHOCState) {
      if (!this.scrollableParents && this.hintRef.current) {
        this.scrollableParents = getScrollableParents(this.hintRef.current);
        this.scrollableParents?.forEach((el) => el.addEventListener('scroll', this.onHide));
      }
      if (prevProps.visible === this.props.visible) return;
      this.changePosition();
    }

    componentWillUnmount() {
      document.removeEventListener('click', this.handleDocumentClick.bind(this));
      window.removeEventListener('scroll', this.onHide);
      window.removeEventListener('resize', this.onHide);
      this.scrollableParents &&
        this.scrollableParents.forEach((el) => el.removeEventListener('scroll', this.onHide));
    }

    handleDocumentClick(e: MouseEvent) {
      if (this.wrapperRef.current && !containNode(this.wrapperRef.current, e.target)) {
        this.onHide();
      }
    }

    changePosition = () => {
      if (!this.props.visible) return;
      const hintNode = this.hintRef.current;
      const wrapperNode = this.wrapperRef.current;
      if (!hintNode || !wrapperNode) return;

      const hintRect = hintNode.getBoundingClientRect();
      const wrapperRect = wrapperNode.getBoundingClientRect();

      const direction: Direction | undefined =
        this.props.direction || findDirection(hintRect, wrapperRect);
      if (!direction) return;

      const align: Align | undefined = this.props.align || findAlign(hintRect, wrapperRect);
      if (!align) return;

      const { parentTop, parentLeft } = getContainingBlockOffset(hintNode);
      const offset = { width: wrapperRect.left - parentLeft, height: wrapperRect.top - parentTop };
      const hintCoords = calculatePosition(direction, align, hintRect, wrapperRect, offset);
      if (!hintCoords) return;

      this.setState({
        pointerAlign: align,
        pointerDirection: direction,
        posX: hintCoords.posX,
        posY: hintCoords.posY,
      });
    };

    handleHide = (e: React.MouseEvent<HTMLDivElement>) => {
      e.stopPropagation();
      this.onHide();
    };

    onHide = () => {
      this.props.onRequestHide?.();
    };

    renderInformer() {
      const { light, message, size = 'big', visible }: IHintHOCProps = this.props;
      const { pointerAlign, pointerDirection, posX, posY } = this.state;

      return (
        <Hint
          light={light}
          offset={this.props.offset}
          pointerAlign={pointerAlign}
          pointerDirection={pointerDirection}
          position={{ posX, posY }}
          ref={this.hintRef}
          visible={visible}
          width={SIZES[size]}
        >
          <CloseIconWrapper light={light} onClick={this.handleHide}>
            <CloseOutline height={20} width={20} />
          </CloseIconWrapper>
          <Body2Long>{message}</Body2Long>
        </Hint>
      );
    }

    render() {
      const { className, container, dataTestId }: IHintHOCProps = this.props;

      return (
        <Wrapper className={className} data-test-id={dataTestId} ref={this.wrapperRef}>
          <WrappedComponent {...this.props} />
          {container
            ? ReactDOM.createPortal(this.renderInformer(), container)
            : this.renderInformer()}
        </Wrapper>
      );
    }
  };
