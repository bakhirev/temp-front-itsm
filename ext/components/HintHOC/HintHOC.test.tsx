import React, { ReactNode } from 'react';
import { mount, shallow } from 'enzyme';

import { Body2Long } from '../Typography';
import { createHintHOC } from '../HintHOC';
import type { IHintHOCProps } from '../HintHOC';

import { Hint } from './Hint';
import { CloseIconWrapper } from './CloseIconWrapper';
import { SIZES } from './constants';

type Align = IHintHOCProps['align'];
type Direction = IHintHOCProps['direction'];
type Size = IHintHOCProps['size'];

jest.mock('../Icons/service/Ð¡loseOutline.svg', () => ({
  // eslint-disable-next-line react/display-name
  ReactComponent: () => <div>x</div>,
}));

jest.mock('../common', () => ({
  DEFAULT_THEME: {
    borderRadius: '',
    color: {
      neutral: {
        80: '',
        white: '',
      },
      primary: {
        40: '',
      },
    },
  },
  Z_INDEXES: {
    HINT: 777,
  },
  markerStyle: () => 'marker style',
}));

jest.mock('../Typography', () => ({
  // eslint-disable-next-line react/display-name
  Body2Long: ({ children }: { children: ReactNode }) => <div>{children}</div>,
}));

describe('HintHOC', () => {
  // jsdom defines the window innerWidth and innerHeight to be 1024 x 768
  const viewPortHeight = window.innerHeight;
  const viewPortWidth = window.innerWidth;

  const wrapperHeight = 20;
  const wrapperLeft = 30;
  const wrapperWidth = 100;

  const hintHeight = 90;
  const hintWidth = 120;

  const positions = {
    bottom: {
      left: { posX: -44, posY: 32 },
      center: { posX: -10, posY: 32 },
      right: { posX: 24, posY: 32 },
    },
    top: {
      left: { posX: -44, posY: -102 },
      center: { posX: -10, posY: -102 },
      right: { posX: 24, posY: -102 },
    },
  };

  const WrappedComponent = () => <div />;
  const WrappedComponentWithHint = createHintHOC(WrappedComponent);
  const mountWrappedComponentWithHint = (
    direction?: Direction,
    align?: Align,
    visible?: boolean,
    onRequestHide?: () => void
  ) =>
    mount(
      <WrappedComponentWithHint
        align={align}
        direction={direction}
        message=""
        visible={visible}
        onRequestHide={onRequestHide}
      />
    );

  const createGetBoundingClientRectMockFn = ({
    bottom = 0,
    height = 0,
    left = 0,
    right = 0,
    top = 0,
    width = 0,
  }: {
    bottom?: number;
    left?: number;
    right?: number;
    top?: number;
    width?: number;
    height?: number;
  }) => (): DOMRect => ({
    bottom,
    left,
    right,
    top,
    width,
    height,
    x: left,
    y: top,
    toJSON: () => {
      // do nothing
    },
  });

  it('should render component', () => {
    const wrapper = shallow(<WrappedComponentWithHint message="" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render wrapped component', () => {
    const wrapper = shallow(<WrappedComponentWithHint message="" />);
    expect(wrapper.find(WrappedComponent)).toHaveLength(1);
  });

  it('should pass all props to wrapped component', () => {
    const props = {
      bar: 'bar',
      foo: 'foo',
      message: '',
    };
    const wrapper = shallow(<WrappedComponentWithHint {...props} />);
    const wrappedComponent = wrapper.find(WrappedComponent);
    expect(wrappedComponent.props()).toEqual(props);
  });

  it('should render hint with provided text using typography component', () => {
    const hintText = 'foo bar baz';
    const wrapper = shallow(<WrappedComponentWithHint message={hintText} />);
    const styledComponent = wrapper.find(Body2Long);
    expect(styledComponent.prop('children')).toEqual(hintText);
  });

  it('should render hint with provided size', () => {
    const hintText =
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.';
    const sizes: Array<Size> = [undefined, 'big', 'medium', 'small'];
    sizes.forEach((size: Size) => {
      const wrapper = shallow(<WrappedComponentWithHint size={size} message={hintText} />);
      const hint = wrapper.find(Hint);
      expect(hint.prop('width')).toEqual(size ? SIZES[size] : SIZES['big']);
    });
  });

  it("shouldn't show hint right after component has been rendered", () => {
    const wrapper = mountWrappedComponentWithHint();
    const hint = wrapper.find(Hint);
    expect(hint.prop('visible')).toBeFalsy();
  });

  it('should show hint when mouse has been clicked on wrapped component', () => {
    const wrapper = mountWrappedComponentWithHint();
    wrapper.setProps({ visible: true });
    const hint = wrapper.find(Hint);
    expect(hint.prop('visible')).toBeTruthy();
  });

  it('should call onRequestHide when mouse has been clicked on "Close" icon', () => {
    let onRequestHideCalled = false;
    const wrapper = mountWrappedComponentWithHint('bottom', 'left', true, () => {
      wrapper.setProps({ visible: false });
      onRequestHideCalled = true;
    });
    expect(onRequestHideCalled).toBeFalsy();
    expect(wrapper.find(Hint).prop('visible')).toBeTruthy();

    const closeIcon = wrapper.find(CloseIconWrapper);
    closeIcon.simulate('click');
    expect(onRequestHideCalled).toBeTruthy();
    expect(wrapper.find(Hint).prop('visible')).toBeFalsy();
  });

  it('should render hint in bottom direction/left align by default if direction/align are not provided and there are left and bottom space available', () => {
    const wrapper = mountWrappedComponentWithHint();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      left: hintWidth + wrapperLeft,
      width: wrapperWidth,
      height: wrapperHeight,
    });

    const hintNode = wrapper.find(Hint).getDOMNode();
    hintNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: hintWidth,
      height: hintHeight,
    });

    wrapper.setProps({ visible: true });
    wrapper.update();

    const hint = wrapper.find(Hint);
    const hintPosition = {
      posX: hintWidth + wrapperLeft + positions.bottom.left.posX,
      posY: positions.bottom.left.posY,
    };
    expect(hint.prop('position')).toEqual(hintPosition);
  });

  it('should render hint in bottom direction/right align by default if direction/align are not provided and there are no left space available', () => {
    const wrapper = mountWrappedComponentWithHint();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      left: wrapperLeft,
      width: wrapperWidth,
      height: wrapperHeight,
    });

    const hintNode = wrapper.find(Hint).getDOMNode();
    hintNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: hintWidth,
      height: hintHeight,
    });

    wrapper.setProps({ visible: true });
    wrapper.update();

    const hintPosition = {
      posX: wrapperLeft + positions.bottom.right.posX,
      posY: positions.bottom.right.posY,
    };
    expect(wrapper.find(Hint).prop('position')).toEqual(hintPosition);
  });

  it('should render hint with big width in bottom direction/center align by default if direction/align are not provided in case of there are no left and no right space available', () => {
    const offset = 6;
    const wrapper = mountWrappedComponentWithHint();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      left: (viewPortWidth - wrapperWidth) / 2,
      width: wrapperWidth,
      height: wrapperHeight,
    });

    const bigHintWidth = viewPortWidth - offset;
    const hintNode = wrapper.find(Hint).getDOMNode();
    hintNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: bigHintWidth,
      height: hintHeight,
    });

    wrapper.setProps({ visible: true });
    wrapper.update();

    const hint = wrapper.find(Hint);
    expect(hint.prop('position')).toEqual({
      posX: offset / 2,
      posY: positions.bottom.center.posY,
    });
  });

  it('should render hint in top direction/left align if direction/align is not provided in case of there are no bottom, but left space available', () => {
    const wrapper = mountWrappedComponentWithHint();

    const wrapperNode = wrapper.getDOMNode();
    wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      bottom: viewPortHeight,
      left: hintWidth + wrapperLeft,
      width: wrapperWidth,
      height: wrapperHeight,
    });

    const hintNode = wrapper.find(Hint).getDOMNode();
    hintNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
      width: hintWidth,
      height: hintHeight,
    });

    wrapper.setProps({ visible: true });
    wrapper.update();

    const hintPosition = {
      posX: hintWidth + wrapperLeft + positions.top.left.posX,
      posY: positions.top.left.posY,
    };
    const hint = wrapper.find(Hint);
    expect(hint.prop('position')).toEqual(hintPosition);
  });

  it('should render hint with or without provided direction and align', () => {
    const directions: Array<Direction> = [undefined, 'bottom', 'top'];
    const aligns: Array<Align> = [undefined, 'left', 'center', 'right'];
    directions.forEach((direction) => {
      aligns.forEach((align) => {
        const wrapper = mountWrappedComponentWithHint(direction, align);

        const wrapperNode = wrapper.getDOMNode();
        wrapperNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
          height: wrapperHeight,
          width: wrapperWidth,
        });

        const hintNode = wrapper.find(Hint).getDOMNode();
        hintNode.getBoundingClientRect = createGetBoundingClientRectMockFn({
          width: hintWidth,
          height: hintHeight,
        });

        wrapper.setProps({ visible: true });
        wrapper.update();

        const hintPosition = positions[direction ?? 'bottom'][align ?? 'right'];
        const hint = wrapper.find(Hint);
        expect(hint.prop('position')).toEqual(hintPosition);
      });
    });
  });
});
