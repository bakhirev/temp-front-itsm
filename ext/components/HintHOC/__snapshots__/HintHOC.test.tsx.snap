// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HintHOC should render component 1`] = `
<styled.div>
  <WrappedComponent
    message=""
  />
  <styled.div
    pointerAlign="left"
    pointerDirection="bottom"
    position={
      Object {
        "posX": 0,
        "posY": 0,
      }
    }
    theme={
      Object {
        "borderRadius": "",
        "color": Object {
          "neutral": Object {
            "80": "",
            "white": "",
          },
          "primary": Object {
            "40": "",
          },
        },
      }
    }
    width={488}
  >
    <styled.div
      onClick={[Function]}
      theme={
        Object {
          "borderRadius": "",
          "color": Object {
            "neutral": Object {
              "80": "",
              "white": "",
            },
            "primary": Object {
              "40": "",
            },
          },
        }
      }
    >
      <ReactComponent
        height={20}
        width={20}
      />
    </styled.div>
    <Body2Long />
  </styled.div>
</styled.div>
`;
